modules/stream-shredder/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/shredder/stream/Resources.scala

```diff
@@ -0,0 +1,65 @@
+package com.snowplowanalytics.snowplow.rdbloader.shredder.stream
+
+import java.util.UUID
+
+import scala.concurrent.duration._
+
+import cats.data.EitherT
+import cats.implicits._
+
+import cats.effect.concurrent.Ref
+import cats.effect.{Blocker, Clock, Resource, Timer, Concurrent, Sync}
+
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+import fs2.concurrent.SignallingRef
+
+import io.circe.Json
+
+import com.snowplowanalytics.iglu.client.Client
+import com.snowplowanalytics.iglu.client.resolver.{InitSchemaCache, InitListCache}
+
+import com.snowplowanalytics.snowplow.rdbloader.common.transformation.EventUtils
+
+case class Resources[F[_]](iglu: Client[F, Json],
+                           atomicLengths: Map[String, Int],
+                           instanceId: String,
+                           blocker: Blocker,
+                           halt: SignallingRef[F, Boolean],
+                           windows: State.Windows[F],
+                           global: Ref[F, Long])
+
+object Resources {
+
+  private implicit def logger[F[_]: Sync] = Slf4jLogger.getLogger[F]
+
+  def mk[F[_]: Concurrent: Clock: InitSchemaCache: InitListCache: Timer](igluConfig: Json): Resource[F, Resources[F]] = {
+    val init = for {
+      igluClient <- Client.parseDefault[F](igluConfig)
+        .leftMap(e => new RuntimeException(s"Error while parsing Iglu config: ${e.getMessage()}"))
+      atomicLengths <- EitherT(EventUtils.getAtomicLengths(igluClient.resolver))
+    } yield (igluClient, atomicLengths)
+    val client = init.value.flatMap {
+      case Right(init) => Sync[F].pure(init)
+      case Left(error) => Sync[F].raiseError[(Client[F, Json], Map[String, Int])](error)
+    }
+
+    for {
+      (client, lengths) <- Resource.liftF(client)
+      blocker <- Blocker[F]
+      state <- Resource.make(State.init[F]) { global =>
+        global.get.flatMap(stack => logger.warn(s"Final window state: $stack"))
+      }
+      sinks <- Resource.liftF(Ref.of(0L))
+      instanceId <- Resource
+        .liftF(Sync[F].delay(UUID.randomUUID()))
+        .evalTap(id => logger[F].info(s"Instantiated $id shredder instance"))
+      halt <- Resource.make(SignallingRef(false)) { s =>
+        logger[F].warn("Halting the source, sleeping for 5 seconds...") *>
+          s.set(true) *>
+          Timer[F].sleep(5.seconds) *>
+          logger[F].warn(s"Shutting down $instanceId instance")
+      }
```

Comment:
From what I've seen in tests this part (`s.set(true)`) is exactly what kills the stream (except abnormal termination). On the other hand, this shutdown mechanism doesn't attempt to do anything with checkpointing which makes it suboptimal.

Reply:
You could test this by (temporarily) putting some logging into the `Processing` stream:

```scala
    getSource[F](resources, config.input)
      .interruptWhen(resources.halt)
      .noneTerminate
      .evalTap {
        case Some(_) => Sync[F].unit
        case None => Sync[F].delay(logger.info("stream was terminated")).void
      }
      .unNoneTerminate
      .through(windowing)
      .evalTap(State.add(resources.windows))
      .through(shred[F](resources.iglu, isTabular, resources.atomicLengths))
      .through(getSink[F](resources.blocker, resources.instanceId, config.output, sinkId, onComplete))
      .flatMap(_.stream)  // Sinks must be issued sequentially
      .compile
      .drain
```

If you send a SIGTERM, I suspect you will never see the message "stream was terminated".  I suspect the IO platform kills the stream first, and then later invokes the release on the resource.

If you *do* see the log message "stream was terminated" then my suspicions are wrong and everything is OK.

Reply:
Ah, I see now where the confusion is coming from. I don't expect anything below your `evalTap` to get new elements, but I do expect that fibers started in `flatMap(_.stream)` will have time to complete gracefully.

In staging what I'm seeing is that without `interruptWhen` it exists immediately at any step, pretty much like `SIGKILL`. With `interruptWhen` I see: if any sink is at the processing it can manage to finish, KCL messages saying its shutting down etc. So at least it's semi-graceful.